### ISSUE TITLE
GetMessage() stuck and timeout

### ISSUE THREAD

**Issue by uhfath** [jstedfast/MailKit#361]
I'm using version from NuGet (1.2.24).
Code is simple enough:
```c#
using (ImapClient client = new ImapClient(new ProtocolLogger("e:\\imap.log")))
{
    client.Connect(Server, Port, true);
    Logger.LogInfo("Connected...");

    client.AuthenticationMechanisms.Remove("XOAUTH2");
    client.Authenticate(Account, Password);
    Logger.LogInfo("Authorized...");

    var inbox = client.Inbox;
    inbox.Open(FolderAccess.ReadWrite);

    if (inbox.Count > 0)
    {
        Logger.LogInfo($"Found messages: {inbox.Count}");

        for (int i = 0; i < inbox.Count;)
        {
            Logger.LogInfo($"Downloading message #{i + 1}");
            var message = inbox.GetMessage(i); //stuck here
            //some processing here
        }
    }
}
```
After some time the exception is thrown. Unfortunately it is localized (russian in my case), so i'm not sure it will be of any help. But here is a callstack:

>    в System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size,
> SocketFlags socketFlags)
>    в System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 si
> ze)
>    --- Конец трассировки внутреннего стека исключений ---
>    в System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 si
> ze)
>    в System.Net.FixedSizeReader.ReadPacket(Byte[] buffer, Int32 offset, Int32 co
> unt)
>    в System.Net.Security._SslStream.StartFrameHeader(Byte[] buffer, Int32 offset
> , Int32 count, AsyncProtocolRequest asyncRequest)
>    в System.Net.Security._SslStream.StartReading(Byte[] buffer, Int32 offset, In
> t32 count, AsyncProtocolRequest asyncRequest)
>    в System.Net.Security._SslStream.ProcessRead(Byte[] buffer, Int32 offset, Int
> 32 count, AsyncProtocolRequest asyncRequest)
>    в System.Net.Security.SslStream.Read(Byte[] buffer, Int32 offset, Int32 count
> )
>    в MailKit.Net.Imap.ImapStream.ReadAhead(Int32 atleast, CancellationToken canc
> ellationToken)
>    в MailKit.Net.Imap.ImapStream.Read(Byte[] buffer, Int32 offset, Int32 count,
> CancellationToken cancellationToken)
>    в MailKit.Net.Imap.ImapFolder.FetchStream(ImapEngine engine, ImapCommand ic,
> Int32 index)
>    в MailKit.Net.Imap.ImapEngine.ProcessUntaggedResponse(CancellationToken cance
> llationToken)
>    в MailKit.Net.Imap.ImapCommand.Step()
>    в MailKit.Net.Imap.ImapEngine.Iterate()
>    в MailKit.Net.Imap.ImapEngine.Wait(ImapCommand ic)
>    в MailKit.Net.Imap.ImapFolder.GetMessage(Int32 index, CancellationToken cance
> llationToken, ITransferProgress progress)
> 
The exception itself says something about timeout and inability to read from socket, because the other endpoint didn't respond in timely fasion.

And so the code SOMETIMES stucks at inbox.GetMessage(i) line.
Sometimes it works. For example i can download 1 or 2 messages or even 3, and then get stucked.
First thought was that connection is dropped. But checked with other clients and with browser - they work OK.
Checked with another library (for instance ImapX) and it also worked OK (but it has other issues).
In this test case i have about 13 messages, each with 3-8 attachments, 10-40 kb per attachment.
It got stuck after the second message got downloaded and processed.
Somtimes it stucks after 6 messages processed. Sometimes right in the beginning.

And BTW, i often get exceptions on authentication. Don't know if it's related, but still just in case:

> MailKit.Net.Imap.ImapProtocolException: The IMAP server has unexpectedly disconn
> ected.
>    в MailKit.Net.Imap.ImapStream.ReadAhead(Int32 atleast, CancellationToken canc
> ellationToken)
>    в MailKit.Net.Imap.ImapStream.ReadToken(String specials, CancellationToken ca
> ncellationToken)
>    в MailKit.Net.Imap.ImapStream.ReadToken(CancellationToken cancellationToken)
>    в MailKit.Net.Imap.ImapEngine.ReadToken(CancellationToken cancellationToken)
>    в MailKit.Net.Imap.ImapCommand.Step()
>    в MailKit.Net.Imap.ImapEngine.Iterate()
>    в MailKit.Net.Imap.ImapEngine.Wait(ImapCommand ic)
>    в MailKit.Net.Imap.ImapClient.Authenticate(Encoding encoding, ICredentials cr
> edentials, CancellationToken cancellationToken)
>    в MailKit.MailService.Authenticate(String userName, String password, Cancella
> tionToken cancellationToken)

Is there anything else i can check?
The log file is pretty huge and contains some sensitive info, so i'm not sure i can post it here. But it ends with "S: A00000007 OK completed"

**Comment by jstedfast** [comment]
I'll need the log, but don't post it on github (it'd be public to everyone).

Send it to jeff@xamarin.com

**Comment by uhfath** [comment]
Just sent the file.
Thanks.

понедельник, 20 июня 2016 г. пользователь Jeffrey Stedfast написал:

**Comment by jstedfast** [comment]
Thanks for the log.

Here's what I discovered:

```
C: A00000007 FETCH 2 (BODY.PEEK[])
S: * 2 FETCH (BODY[] {225752}
<182374 bytes>
S: )
S: A00000007 OK completed
```

The reason that MailKit is hanging in this case is that the server told MailKit that it would be sending 225752 bytes of message data but it only sent 182374 bytes of message data, which means that MailKit is still waiting to read another 43378 bytes from the server.

The reason that some other IMAP libraries "work" is because they are broken and do not correctly parse IMAP responses. What they do instead is that they read lines until they read a line that starts with `)` -or- until they read a line that matches `A00000007 OK completed`, both of which are wrong.

The reason that reading until `)` is wrong is probably obvious (there's no reason a message can't have a line starting with a close-parenthesis).

The reason for the second might not be as obvious, but it follows the same logic as the first. For example, what if you send me a log file of IMAP commands in a text attachment (minus the C: and S: prefixes that MailKit uses to make things easier to read) and I use one of these IMAP libraries to read the message and it *just so happens* that the command tag that the IMAP library uses is the same as the one used in the log file? It would totally break the library, wouldn't it? :)

So in summary, the bug is in the IMAP server giving MailKit an incorrect byte count.

**Comment by jstedfast** [comment]
To better illustrate my point about lines beginning with the tag, consider the following maliciously designed email message:

```
From: Evil Hacker <evil-hacker@example.com>
To: You <you@example.com>
Subject: muhahaha, this will break your IMAP client!
Content-Type: text/plain

A00000005 OK completed
A00000006 OK completed
A00000007 OK completed
A00000008 OK completed
A00000009 OK completed
A00000010 OK completed
A00000011 OK completed
...
A00010000 OK completed
```

If MailKit just read until it found the tag it used for the command at the beginning of a line, then it would stop reading way too early and then all of the following commands would simply read 1 more line of data from the message data thinking it was the end of the command and get completely out of whack.

**Comment by uhfath** [comment]
I understand your point.
So the question is, can this library be used with any server out there or just some chosen ones?
For instance, the server i used it with is a very big company and contains millions of users who use it daily. And if i'm about to create a mail client or some other functionality with your library, what should i do in case a user used a bad mail server?
Or perhaps there is a way to make this work with, for example, some setting in code? Which defaults to normal (current) behaviour, but can be opted to work in this case scenario.

**Comment by jstedfast** [comment]
Well, theoretically, MailKit can be used with *all* IMAP servers, but in reality, it can only be used with IMAP servers that properly implement the specification.

I've added numerous work-arounds for broken IMAP servers, but this particular server bug is not possible to work around in any sane foolproof way :(

Unfortunately there's no standard way of toggling an IMAP server into rfc-compliant mode, so if this particular IMAP server has an rfc-compliant mode that you can toggle it into, I'm not sure how you'd do it.

You could try emailing the company that maintains the IMAP server and let them know the bug exists, maybe they'd be willing to fix it?

**Comment by uhfath** [comment]
Yeah, but probably they won't even receive my email (LOL).
And even if they do, they won't be willing to make such changes and possibly break all other clients, that just work.
Thank you for your time.

**Comment by jstedfast** [comment]
Fixing the bug won't break the other clients.

**Comment by jvilhuber** [comment]
@jstedfast Just some random musings here: Is there a reason IMAP commands start with 'A' (I've also seen 'B') and some numbers? While it wouldn't completely address your point about the embedded IMAP command in an email, it would certainly reduce the likelihood of it happening, if MailKit used "<x random characters><y digits>" as the token (for some values of x, say 8, and y, say 5). I would think using some randomization scheme would require keeping more state in the client, so that's not ideal, and certainly a server that sends the wrong size for the data is clearly broken, but there's ways to catch this case, given that the server companies may have little incentive to fix one user's bug.

For example, one could gracefully recover IFF the timeout is reached AND the token is the 'OK Completed' line, especially, if the command identifier is not predictable. Or one could be slightly more aggressive by looking for the "ok completed" line, then waiting some small-ish timeout value after that to see if there's more data (if the data received is less than initially claimed by the server).

Just thinking out loud (so to speak).

**Comment by jstedfast** [comment]
Once you get a timeout, as far as I know, the socket is no longer usable. Even if you *can* continue using it after the TCP timeout, it still means you'll get a hang.

Using a randomly generated command token instead of a sequential token like MailKit currently uses wouldn't be difficult to do, it's just that a sequential tag was simpler to implement since the only "requirement" was that the tag be unique (it doesn't actually have to be unique, but it is helpful if/when MailKit ever gets support for IMAP command pipelining).

The hard part is that the logic to figure out when to stop reading would get ugly and since this bug only exists with 1 server, it doesn't seem worth the effort - especially since it'd probably be better to just report a bug to the server and have them fix their bug (which is probably a 5 minute fix).

**Comment by jstedfast** [comment]
WRT socket timeouts, a possible workaround for that is polling in a timeout like the IMAP logic already does (needed for IDLE support), however that doesn't work for SslStream.
